src/main/kotlin/io/github/inductiveautomation/kindling/utils/Diff.kt

```diff
@@ -0,0 +1,209 @@
+package io.github.inductiveautomation.kindling.utils
+
+import io.github.inductiveautomation.kindling.utils.LongestCommonSequence.Companion.lcs
+import kotlin.math.max
+
+typealias DiffList<T> = List<Diff<T>>
+
+sealed interface Diff<T> {
+    val type: DiffType
+    val value: T
+    val index: Int
+
+    class Addition<T>(
+        override val type: DiffType,
+        override val value: T,
+        override val index: Int,
+    ) : Diff<T>
+
+    class Deletion<T>(
+        override val type: DiffType,
+        override val value: T,
+        override val index: Int,
+    ) : Diff<T>
+
+    class NoChange<T>(
+        override val type: DiffType,
+        override val value: T,
+        val preIndex: Int?,
+        val postIndex: Int?,
+    ) : Diff<T> {
+        override val index: Int
+            get() = (preIndex ?: postIndex)!!
+    }
+
+    companion object {
+        operator fun <T> invoke(
+            type: DiffType,
+            value: T,
+            index1: Int,
+            index2: Int? = null,
+        ): Diff<T> {
+            return when (type) {
+                DiffType.ADDITION -> Addition(type, value, index1)
+                DiffType.DELETION -> Deletion(type, value, index1)
+                DiffType.NOCHANGE -> NoChange(type, value, index1, index2)
+            }
+        }
+    }
+}
+
+@Suppress("MemberVisibilityCanBePrivate")
+class DiffData<T> private constructor(
+    val preChange: List<T>,
+    val postChange: List<T>,
+    val additions: DiffList<T>,
+    val deletions: DiffList<T>,
+) {
+    private constructor(
+        preChange: List<T>,
+        postChange: List<T>,
+        equalityPredicate: (T, T) -> Boolean,
+        lcs: List<T> = lcs(preChange, postChange, equalityPredicate),
+    ) : this(
+        preChange,
+        postChange,
+        additions = postChange.mapIndexedNotNull { index, item ->
+            if (item in lcs) null else Diff(DiffType.ADDITION, item, index)
+        }.sortedBy { it.index },
+        deletions = preChange.mapIndexedNotNull { index, item ->
+            if (item in lcs) null else Diff(DiffType.DELETION, item, index)
+        }.sortedBy { it.index },
+    )
+
+    constructor(
```

Comment:
Telescoping constructors :ew:
Generally, I prefer a factory function or something for this